src/core/data.jl

```diff
@@ -523,3 +541,34 @@ function _make_full_matrix_variable(diag, lowertriangle, uppertriangle)
     # matrix = diagm(0 => diag) + _vec2ltri!(lowertriangle) + _vec2utri!(uppertriangle)
     return matrix
 end
+
+
+function _has_nl_expression(x)::Bool
+    if isa(x, JuMP.NonlinearExpression)
+        return true
+    elseif isa(x, Vector)
+        for i in x
+            if _has_nl_expression(i)
+                return true
+            end
+        end
+    elseif isa(x, Dict)
+        for i in values(x)
+            if _has_nl_expression(i)
+                return true
+            end
+        end
+    end
+    return false
+end
+
+
+macro smart_constraint(model, vars, expr)
+    esc(quote
+        if _has_nl_expression($vars)
+            JuMP.@NLconstraint($model, $expr)
+        else
+            JuMP.@constraint($model, $expr)
+        end
+    end)
+end
```

Comment:
I added a comment about this, but it seemed to have been lost. This has to do with performance improvements in KCL. At this moment, all KCL equations have become NLconstraints. This method would allow you to choose in a smart way whether you need a NLconstraint or not in place. However, adding custom macros seemed controversial to me, so I decided to postpone this until PMD v10